rfcs/20191212-keras-categorical-inputs.md

```diff
@@ -0,0 +1,357 @@
+
+dftrain = pd.read_csv('https://storage.googleapis.com/tf-datasets/titanic/train.csv')
+y_train = dftrain.pop('survived')
+
+dataset = tf.data.Dataset.from_tensor_slices((
+	(tf.to_sparse(dftrain.sex, "Unknown"), tf.to_sparse(dftrain.n_siblings_spouses, -1),
+	tf.to_sparse(dftrain.parch, -1), tf.to_sparse(dftrain['class'], "Unknown"), tf.to_sparse(dftrain.deck, "Unknown"),
+	tf.expand_dims(dftrain.age, axis=1), tf.expand_dims(dftrain.fare, axis=1)),
+	y_train)).batch(bach_size).repeat(n_epochs)
+
+model.fit(dataset)
+```
+
+### Workflow 2
+
+The second example gives an instruction on how to transition from categorical feature columns to the proposed layers. Note that one difference for vocab categorical column is that, instead of providing a pair of mutually exclusive `default_value` and `num_oov_buckets` where `default_value` represents the value to map input to given out-of-vocab value, and `num_oov_buckets` represents value range of [len(vocab), len(vocab)+num_oov_buckets) to map input to from a hashing function given out-of-vocab value. In practice, we believe out-of-vocab values should be mapped to the head, i.e., [0, num_oov_tokens), and in-vocab values should be mapped to [num_oov_tokens, num_oov_tokens+len(vocab)).
+
+1. Categorical vocab list column
+
+Original:
+```python
+fc = tf.feature_column.categorical_feature_column_with_vocabulary_list(
+	   key, vocabulary_list, dtype, default_value, num_oov_buckets)
+```
+Proposed:
+```python
+x = tf.keras.Input(shape=(1,), name=key, dtype=dtype)
+layer = tf.keras.layers.CategoryLookup(
+            vocabulary=vocabulary_list, num_oov_tokens=num_oov_buckets)
+out = layer(x)
+```
+
+2. categorical vocab file column
+
+Original:
+```python
+fc = tf.feature_column.categorical_column_with_vocab_file(
+       key, vocabulary_file, vocabulary_size, dtype,
+       default_value, num_oov_buckets)
+```
+Proposed:
+```python
+x = tf.keras.Input(shape=(1,), name=key, dtype=dtype)
+layer = tf.keras.layers.CategoryLookup(
+            vocabulary=vocabulary_file, num_oov_tokens=num_oov_buckets)
+out = layer(x)
+```
+Note: `vocabulary_size` is only valid if `adapt` is called. Otherwise if user desires to lookup for the first K vocabularies in vocab file, then shrink the vocab file by only having the first K lines.
+
+3. categorical hash column
+
+Original:
+```python
+fc = tf.feature_column.categorical_column_with_hash_bucket(
+       key, hash_bucket_size, dtype)
+```
+Proposed:
+```python
+x = tf.keras.Input(shape=(1,), name=key, dtype=dtype)
+layer = tf.keras.layers.CategoryHashing(num_bins=hash_bucket_size)
+out = layer(x)
+```
+
+4. categorical identity column
+
+Original:
+```python
+fc = tf.feature_column.categorical_column_with_identity(
+       key, num_buckets, default_value)
+```
+Proposed:
+```python
+x = tf.keras.Input(shape=(1,), name=key, dtype=dtype)
+layer = tf.keras.layers.Lambda(lambda x: tf.where(tf.logical_or(x < 0, x > num_buckets), tf.fill(dims=tf.shape(x), value=default_value), x))
```

Comment:
If we allowed to specify the hash function, this could also be folded into the CategoryHashing with an IdentityHash.

Reply:
CategoryHashing does not check oov values, so if we have that then it's complicating the signature.

Reply:
Why do you not use a layer to the lamda layer?

Reply:
How about add a layer to do the Identity Work? 
1. The lambda expression is a little complicated
2. We may need the layer to handle both dense tensor and SparseTensor.